```python
def fetcher(obj, index):
    return obj[index]


x = list(range(2))
try:
    fetcher(x, 2)
except IndexError:
    print('got exception')
print('continuing')

# 手动引发异常 ，没有捕获到就会向上传递
try:
    raise IndexError
except IndexError:
    print("something")

# 有条件的 raise
assert 1 > 0, 'what the hell?'


# 定义异常
class Bad(Exception):
    def __init__(self):
        print('I am bad')

    pass


'''
class doomed():
    raise Bad()


try:
    doomed()
except Bad:
    print('I got Bad')
'''

# finally 细节
'''
try:
    raise IndexError
finally:
    print('I am run')
print('I am not run')
'''

try:
    raise IndexError
except IndexError:
    print('error')
finally:
    print('I am running')
print('I am running too')

# with as  =>  对象的环境管理器  自动关闭文件，对锁的自动上锁和开锁
with open('1.txt', 'w') as file:
    file.write('The larch \n')

# else 是没发声异常时执行的异常处理器
try:
    fetcher(x, 1)
except IndexError:
    print('Index Error')
except KeyError:
    print('KeyError')
else:
    print('nothing happens')

# except 可以列出一组异常 except (e1,e2,e3)
'''
try:
    something...
except:
    catch something...
else:
    not
finally:
    finally catch
'''

# raise from 
try:
    1/0
except Exception as E:
    raise TypeError('Bad') from E



```

